cli/cmd/completion.go

```diff
@@ -0,0 +1,251 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"io"
+	"os"
+
+	"github.com/spf13/cobra"
+	"helm.sh/helm/v3/cmd/helm/require"
+)
+
+const completionDesc = `
+Generate autocompletions script for lokoctl for the specified shell (bash or zsh).
+`
+
+const bashCompDesc = "\nGenerate the autocompletion script for lokoctl for the bash shell. \n\n" +
+
+	"To load completions in your current shell session: \n" +
+	"`$ source <(lokoctl completion bash)` \n\n" +
+
+	"To load completions for every new session, execute once: \n" +
+	"Linux: `$ lokoctl completion bash > /etc/bash_completion.d/lokoctl && exec $SHELL` \n" +
+	"MacOS: `$ lokoctl completion bash > /usr/local/etc/bash_completion.d/lokoctl && exec $SHELL` "
+
+const zshCompDesc = "\nGenerate the autocompletion script for lokoctl for the zsh shell. \n\n" +
+
+	"To load completions in your current shell session: \n" +
+	"`$ source <(lokoctl completion zsh)` \n\n" +
+
+	"To load completions for every new session, execute once: \n" +
+	"`$ lokoctl completion zsh > \"${fpath[1]}/_lokoctl\" && exec $SHELL`"
+
+func newCompletionCmd(out io.Writer) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:               "completion",
+		Short:             "Generate autocompletions script for the specified shell",
+		Long:              completionDesc,
+		Args:              require.NoArgs,
+		ValidArgsFunction: noCompletions, // Disable file completion
+	}
+
+	bash := &cobra.Command{
+		Use:                   "bash",
+		Short:                 "Generate autocompletions script for bash",
+		Long:                  bashCompDesc,
+		Args:                  require.NoArgs,
+		DisableFlagsInUseLine: true,
+		ValidArgsFunction:     noCompletions,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return runCompletionBash(out, cmd)
+		},
+	}
+
+	zsh := &cobra.Command{
+		Use:                   "zsh",
+		Short:                 "Generate autocompletions script for zsh",
+		Long:                  zshCompDesc,
+		Args:                  require.NoArgs,
+		DisableFlagsInUseLine: true,
+		ValidArgsFunction:     noCompletions,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return runCompletionZsh(out, cmd)
+		},
+	}
+
+	cmd.AddCommand(bash, zsh)
+
+	return cmd
+}
+
+func runCompletionBash(out io.Writer, cmd *cobra.Command) error {
+	return cmd.Root().GenBashCompletion(out)
+}
+
+func runCompletionZsh(out io.Writer, cmd *cobra.Command) error { //nolint:funlen
```

Comment:
Why not use `GenZshCompletion()` completion for zsh?

Reply:
I was using that before, but it flags were preferred over nouns. https://github.com/spf13/cobra/issues/1189

Reply:
Cobra 1.0.0 does not support `ValidArgsFunction` for `zsh`.  You will need to wait for the next release, or you can point to the master branch which has the full support required.

Reply:
@knrt10 what's the diff between the version generated by cobra and this one? Can we maybe patch it dynamically to avoid hardcoding it?

Reply:
> what's the diff between the version generated by cobra and this one? Can we maybe patch it dynamically to avoid hardcoding it?

If you allow me to answer this one.
On cobra's master branch, there was a complete redesign of the zsh completion solution.  
It was a major change as you can see from:
https://github.com/spf13/cobra/pull/1070

Until it is released, the zsh completion (of cobra 1.0) is incomplete.

The alternative used in this pr is what `helm` and `kubectl` currently use for zsh, until the next cobra release.  This temporary solution is based on bash completion which is fully supported in cobra 1.0.

Reply:
@marckhouzam thanks for the reply! And when do you plan 1.0 release roughly?

Reply:
@invidian I'm not a maintainer of the Cobra project but the suggestion of a new release has been mentioned. I'm hopeful it will happen in the next few weeks, but no guarantee.